---
title: "phylosophy"
author:
-   name: Vinh Tran
    affiliation: Goethe University Frankfurt, Frankfurt am Main, Germany
    email: tran@bio.uni-frankfurt.de
date: "`r Sys.Date()`"
output:
    BiocStyle::html_document:
        toc_float: true
        number_sections: false
package: phylosophy
abstract: |
    phylosophy: a comprehensive toolkit for developing and exploring complex 
    phylogenetic profiles
vignette: >
    %\VignetteIndexEntry{phylosophy}
    %\VignetteEngine{knitr::rmarkdown}
    %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
    library(phylosophy),
    collapse = TRUE,
    comment = "#>",
    dev = 'png',
    crop = NULL
)
```

# Introduction
Phylogenetic profiles capture the presence - absence pattern of genes across
species (Pellegrini et al., 1999). The presence of an ortholog in a given
species is often taken as evidence that also the corresponding function is
represented (Lee et al., 2007). Moreover, if two genes agree in their
phylogenetic profile, it can suggest that they functionally interact
(Pellegrini et al., 1999). Phylogenetic profiles are therefore commonly used
for tracing functional protein clusters or metabolic networks across species
and through time. However, orthology inference is not error-free (Altenhoff
et al., 2016), and orthology does not guarantee functional equivalence for
two genes (Studer and Robinson-Rechavi, 2009). Therefore, phylogenetic
profiles are often integrated with accessory information layers, such as
sequence similarity, domain architecture similarity, or semantic similarity
of Gene Ontology-term descriptions.

Various approaches exist to visualize such profiles. However, there is still a
shortage of tools that provide a comprehensive set of functions for the display,
filtering and analysis of multi-layered phylogenetic profiles comprising
hundreds of genes and taxa. To close this methodological gap, we present here
**PhyloProfile**, an *R-based tool to visualize, explore and analyze
multi-layered phylogenetic profiles*.

# How to install phylosophy

Install *phylosophy* from [github](https://github.com/BIONF/phylosophy):

```r
if (!requireNamespace("devtools"))
    install.packages("devtools")
devtools::install_github("BIONF/phylosophy", INSTALL_opts = c('--no-lock'), build_vignettes = TRUE)
```

# Input


# Features and capabilities

# Examples
## Process raw input


**More examples? Please tell us what you want to see ;-)**

# How to cite
>Ngoc-Vinh Tran, Bastian Greshake Tzovaras, Ingo Ebersberger, PhyloProfile: 
dynamic visualization and exploration of multi-layered phylogenetic profiles, 
Bioinformatics, Volume 34, Issue 17, 01 September 2018, Pages 3041‚Äì3043, 
https://doi.org/10.1093/bioinformatics/bty225

Or use the citation function in R CMD to have the citation in BibTex or
LaTeX format

```{r}
citation("PhyloProfile")
```

# How to contribute
![](http://www.sharegif.com/wp-content/uploads/2014/03/giphy41.gif)

Thanks so much for your interest in contributing to *PhyloProfile* üéâüëçüçæ

Contributions to *PhyloProfile* can take many forms. If you are

**biologist**, you can

- [report bugs](https://github.com/BIONF/PhyloProfile/issues/new) for both
online & standalone version,
- [tell us about what features you would love to see](https://goo.gl/SBU4EG),
- improve our documentation, both
[in the Wiki](https://github.com/BIONF/PhyloProfile/wiki) and in our [R
ADME](https://github.com/BIONF/PhyloProfile/blob/master/README.md),
- discuss about [non-coding issues](https://goo.gl/bDS9Cb)

**biologist and love coding**, you can

- [fix existing bugs](https://github.com/BIONF/PhyloProfile/issues/),
- or add new features. Some things we'd love to see are: add scripts to add
out-of-the-box support for further orthology prediction tools, increasing the
test coverage from 0% to something above that, or basically whatever great
idea you have!
- and all points for non-coding contributors as well :)

**not biologist but can code**, it would be great if you can

- test the tool in different environments (Windows, Linux, Mac - Firefox,
Chrome, IE, Safari,...),
- suggest a better user interface,
- improve the code quality

Don't hesitate to get in touch with us if you have any questions. You can
contact us at [tran@bio.uni-frankfurt.de](mailto:tran@bio.uni-frankfurt.de)

## Contributors {.unnumbered}
- [Vinh Tran](https://github.com/trvinh)

# SessionInfo()

Here is the output of `sessionInfo()` on the system on which this document was
compiles:

```{r sessionInfo, echo = FALSE}
# sessionInfo(package = "phylosophy")
```

# References

1. Adebali, O. and Zhulin, I.B. (2017) Aquerium: A web application for
comparative exploration of domain-based protein occurrences on the taxonomically
clustered genome tree. Proteins, 85, 72-77.

2. Altenhoff, A.M. et al. (2016) Standardized benchmarking in the quest for
orthologs. Nat Methods, 13, 425-430.

3. Altenhoff, A.M. et al. (2015) The OMA orthology database in 2015: function
predictions, better plant support, synteny view and other improvements.
Nucleic Acids Res, 43, D240-249.

4. Capra, J.A. et al. (2013) How old is my gene? Trends Genet, 29, 659-668.
Daubin, V., Gouy, M. and Perriere, G. (2002) A phylogenomic approach to
bacterial phylogeny: evidence of a core of genes sharing a common history.
Genome Res, 12, 1080-1090.

5. Huerta-Cepas, J., Serra, F. and Bork, P. (2016) ETE 3: Reconstruction,
Analysis, and Visualization of Phylogenomic Data. Mol Biol Evol, 33, 1635-1638.

6. Koestler, T., Haeseler, A.v. and Ebersberger, I. (2010) FACT: Functional
annotation transfer between proteins with similar feature architectures.
BMC Bioinformatics, 11, 417.

7. Lee, D., Redfern, O. and Orengo, C. (2007) Predicting protein function
from sequence and structure. Nat Rev Mol Cell Biol, 8, 995-1005.

8. Moore, A.D. et al. (2014) DoMosaics: software for domain arrangement
visualization and domain-centric analysis of proteins. Bioinformatics,
30, 282-283.

9. Pellegrini, M. (2012) Using phylogenetic profiles to predict functional
relationships. Methods Mol Biol, 804, 167-177.

10. Pellegrini, M. et al. (1999) Assigning protein functions by comparative
genome analysis: protein phylogenetic profiles. Proc Natl Acad Sci U S A, 96,
4285-4288.

11. Schmitt, T. et al. (2011) Letter to the editor: SeqXML and OrthoXML:
standards for sequence and orthology information. Brief. Bioinform., 12,
485-488.

12. Studer, R.A. and Robinson-Rechavi, M. (2009) How confident can we be that
orthologs are similar, but paralogs differ? Trends Genet, 25, 210-216.
